libethcore/EthashAux.cpp

```diff
@@ -240,8 +240,9 @@ Ethash::Result EthashAux::eval(BlockInfo const& _header, Nonce const& _nonce)
 
 Ethash::Result EthashAux::eval(h256 const& _seedHash, h256 const& _headerHash, Nonce const& _nonce)
 {
-	if (FullType dag = get()->m_fulls[_seedHash].lock())
-		return dag->compute(_headerHash, _nonce);
+	DEV_GUARDED(get()->x_fulls)
+		if (FullType dag = get()->m_fulls[_seedHash].lock())
```

Comment:
This method seems a bit lopsided. Why is it checking m_full only to return the value from light, if it's not found?  

Also, I think light() maybe a bit messy as well. This line:
LightType ret = get()->m_lights[_seedHash];

Will always return a value. May be some other bugs lurking.